webapp/channels/src/components/advanced_text_editor/use_key_handler.tsx

```diff
@@ -0,0 +1,368 @@
+    handleDraftChange: (draft: PostDraft, options?: {instant?: boolean; show?: boolean}) => void,
+    handleSubmit: (e: React.FormEvent, submittingDraft?: PostDraft) => void,
+    emitTypingEvent: () => void,
+    toggleShowPreview: () => void,
+    toggleAdvanceTextEditor: () => void,
+    toggleEmojiPicker: () => void,
+    replyToLastPost: ((e: React.KeyboardEvent) => void) | undefined,
+): [
+        (e: React.KeyboardEvent<TextboxElement>) => void,
+        (e: React.KeyboardEvent<TextboxElement>) => void,
+    ] => {
+    const dispatch = useDispatch();
+
+    const ctrlSend = useSelector((state: GlobalState) => getBool(state, Preferences.CATEGORY_ADVANCED_SETTINGS, 'send_on_ctrl_enter'));
+    const codeBlockOnCtrlEnter = useSelector((state: GlobalState) => getBool(state, Preferences.CATEGORY_ADVANCED_SETTINGS, 'code_block_ctrl_enter', true));
+    const messageHistory = useSelector((state: GlobalState) => state.entities.posts.messagesHistory.messages);
+
+    const timeoutId = useRef<number>();
+    const messageHistoryIndex = useRef(messageHistory.length);
+    const lastChannelSwitchAt = useRef(0);
+    const isNonFormattedPaste = useRef(false);
+
+    const onEditLatestPost = useCallback((e: React.KeyboardEvent) => {
+        e.preventDefault();
+        const {data: canEditNow} = dispatch(editLatestPost(channelId, postId));
+        if (!canEditNow) {
+            focusTextbox(true);
+        }
+    }, [focusTextbox, channelId, postId, dispatch]);
+
+    const loadPrevMessage = useCallback((e: React.KeyboardEvent) => {
+        e.preventDefault();
+        if (messageHistoryIndex.current === 0) {
+            return;
+        }
+        messageHistoryIndex.current -= 1;
+        handleDraftChange({
+            ...draft,
+            message: messageHistory[messageHistoryIndex.current] || '',
+        });
+    }, [draft, handleDraftChange, messageHistory]);
+
+    const loadNextMessage = useCallback((e: React.KeyboardEvent) => {
+        e.preventDefault();
+        if (messageHistoryIndex.current >= messageHistory.length) {
+            return;
+        }
+        messageHistoryIndex.current += 1;
+        handleDraftChange({
+            ...draft,
+            message: messageHistory[messageHistoryIndex.current] || '',
+        });
+    }, [draft, handleDraftChange, messageHistory]);
+
+    const postMsgKeyPress = useCallback((e: React.KeyboardEvent<TextboxElement>) => {
+        const {allowSending, withClosedCodeBlock, ignoreKeyPress, message} = postMessageOnKeyPress(
+            e,
+            draft.message,
+            ctrlSend,
+            codeBlockOnCtrlEnter,
+            postId ? 0 : Date.now(),
+            postId ? 0 : lastChannelSwitchAt.current,
+            caretPosition,
+        );
+
+        if (ignoreKeyPress) {
+            e.preventDefault();
+            e.stopPropagation();
+            return;
+        }
+
+        if (allowSending && isValidPersistentNotifications) {
```

Comment:
I imagine your work would've be a lot easier if we had a better way of structuring things that can prevent the post from being sent instead of them being handled differently in different functions. For example, I don't know why `isValidPersistentNotifications` isn't part of `allowSending`

Reply:
Originally `isValidPersistentNotifications` was a function.

On top of that, they are two different features added at different times, so probably there is some "history reasons".

Finally, this allowSending considers the event and the message, not the complete post (i.e. whether it has set the persistent notifications metadata). We could pass the complete draft? sure. But we are not as of today.